File: baekjoon/스택_큐/괄호_9012.py
n = int(input())
for _ in range(n):
    case = str(input())
    if len(case) % 2 != 0:
        print("NO")
        continue
    elif case[-1] == "(" or case[0] == ")":
        print("NO")
        continue

    stack = []
    is_valid = True
    for c in case:
        if c == "(":
            stack.append(c)
        elif stack != [] and c == ")":
            stack.pop()
        elif stack == [] and c == ")":
            is_valid = False
            break

    if is_valid == True and len(stack) == 0:
        print("YES")
    else:
        print('NO')

# 2회차 풀이 2024-10-26
# 괄호 9012번 2회차 풀이
# 스택
n = int(input())
for _ in range(n):
    ps = str(input())
    stack = [] # (만 들어가는 스택
    flag = True
    for p in ps:
        if p == "(":
            stack.append(p)
        else:
            if stack == []:
                print("NO")
                flag = False
                break
            else:
                stack.pop()
    if flag:
        if len(stack) > 0:
            print("NO")
        elif stack == []:
            print("YES")

File: baekjoon/스택_큐/괄호_9012.js
function solution(s) {
    const stack = [];
    for (const i of s) {
        if (i === "(") {
            stack.push(i);
        } else {
            if (stack.length > 0) {
                stack.pop();
            } else {
                return "NO";
            }
        }
    }
    return stack.length === 0 ? "YES" : "NO";
}

const input = require("fs")
    .readFileSync("/dev/stdin")
    .toString()
    .trim()
    .split("\n");
const n = Number(input[0]);
for (let i = 1; i <= n; i++) {
    const s = input[i];
    console.log(solution(s));
}
